From f62ff048443c9752f573457808a7919fb5b5cdcd Mon Sep 17 00:00:00 2001
From: Jiacheng Zheng <jiaczhen@codeaurora.org>
Date: Fri, 29 Jan 2021 15:50:05 +0800
Subject: [PATCH] parallelize audio and pcie driver to improve bootkpi

Audio driver takes the most booting time, so we pick
out all audio dependency chain and run them in a
separate thread. Pcie init also takes some time and we
can define it as asynchronous to save more time.
Finally, we postpone the calling of wait_for_device_probe
after the mounting of rootfs so considerable rootfs mounting
time can be hidden in parallel with audio driver init.

Change-Id: I08f1e8c7089afc55acc4236051e3b0e2594c983d
Signed-off-by: Jiacheng Zheng <jiaczhen@codeaurora.org>
---
 drivers/base/dd.c                                  |   8 +-
 drivers/pci/host/pci-msm.c                         |   1 +
 drivers/soc/qcom/qdsp6v2/apr_vm.c                  |  15 +--
 drivers/soc/qcom/qdsp6v2/msm_audio_ion_vm.c        |   5 +-
 drivers/soc/qcom/qdsp6v2/voice_svc.c               |   5 +-
 drivers/staging/android/ion/ion_hvenv_driver.c     |   5 +-
 init/do_mounts.c                                   |   3 +-
 sound/core/compress_offload.c                      |   5 +-
 sound/core/pcm.c                                   |   5 +-
 sound/core/rawmidi.c                               |   5 +-
 sound/core/timer.c                                 |   5 +-
 sound/last.c                                       |   5 +-
 sound/soc/codecs/msm_stub.c                        |   5 +-
 sound/soc/msm/apq8096-auto.c                       | 109 ++++++++++++++++++++-
 sound/soc/msm/msm-dai-fe.c                         |   5 +-
 sound/soc/msm/msm-pcm-hostless.c                   |   5 +-
 sound/soc/msm/qdsp6v2/audio_slimslave.c            |   5 +-
 sound/soc/msm/qdsp6v2/msm-compress-q6-v2.c         |   5 +-
 sound/soc/msm/qdsp6v2/msm-dai-q6-hdmi-v2.c         |   5 +-
 sound/soc/msm/qdsp6v2/msm-dai-q6-v2.c              |   5 +-
 sound/soc/msm/qdsp6v2/msm-dai-slim.c               |   5 +-
 sound/soc/msm/qdsp6v2/msm-dai-stub-v2.c            |   5 +-
 sound/soc/msm/qdsp6v2/msm-lsm-client.c             |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-afe-v2.c             |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-dtmf-v2.c            |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-host-voice-v2.c      |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-loopback-v2.c        |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-q6-noirq.c           |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-q6-v2.c              |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-routing-v2.c         |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-voice-v2.c           |   5 +-
 sound/soc/msm/qdsp6v2/msm-pcm-voip-v2.c            |   5 +-
 .../soc/msm/qdsp6v2/msm-transcode-loopback-q6-v2.c |   5 +-
 sound/soc/msm/qdsp6v2/q6adm.c                      |   5 +-
 sound/soc/msm/qdsp6v2/q6afe.c                      |   5 +-
 sound/soc/msm/qdsp6v2/q6asm.c                      |   5 +-
 sound/soc/msm/qdsp6v2/q6core.c                     |   5 +-
 sound/soc/msm/qdsp6v2/q6lsm.c                      |   5 +-
 sound/soc/msm/qdsp6v2/q6voice.c                    |   5 +-
 sound/soc/msm/qdsp6v2/rtac.c                       |   5 +-
 sound/soc/soc-core.c                               |   5 +-
 41 files changed, 231 insertions(+), 85 deletions(-)

diff --git a/drivers/base/dd.c b/drivers/base/dd.c
index 0ab3a8d..2186d95 100644
--- a/drivers/base/dd.c
+++ b/drivers/base/dd.c
@@ -666,14 +666,14 @@ static int __driver_attach(struct device *dev, void *data)
 
 	if (!driver_match_device(drv, dev))
 		return 0;
-	if (lock_parent(dev))
-		device_lock(dev->parent);
+	//if (lock_parent(dev))
+		//device_lock(dev->parent);
 	device_lock(dev);
 	if (!dev->driver)
 		driver_probe_device(drv, dev);
 	device_unlock(dev);
-	if (lock_parent(dev))
-		device_unlock(dev->parent);
+	//if (lock_parent(dev))
+		//device_unlock(dev->parent);
 
 	return 0;
 }
diff --git a/drivers/pci/host/pci-msm.c b/drivers/pci/host/pci-msm.c
index 9be5d60..7d44189 100644
--- a/drivers/pci/host/pci-msm.c
+++ b/drivers/pci/host/pci-msm.c
@@ -6638,6 +6638,7 @@ static struct platform_driver msm_pcie_driver = {
 	.driver	= {
 		.name		= "pci-msm",
 		.owner		= THIS_MODULE,
+		.probe_type = PROBE_PREFER_ASYNCHRONOUS,
 		.of_match_table	= msm_pcie_match,
 	},
 };
diff --git a/drivers/soc/qcom/qdsp6v2/apr_vm.c b/drivers/soc/qcom/qdsp6v2/apr_vm.c
index 281d127..521663b 100644
--- a/drivers/soc/qcom/qdsp6v2/apr_vm.c
+++ b/drivers/soc/qcom/qdsp6v2/apr_vm.c
@@ -1155,7 +1155,8 @@ static void apr_vm_set_subsys_state(void)
 	apr_set_modem_state(APR_SUBSYS_LOADED);
 }
 
-static int __init apr_init(void)
+//static int __init apr_init(void)
+int __init apr_init(void)
 {
 	int i, j, k;
 	int ret;
@@ -1229,9 +1230,10 @@ static int __init apr_init(void)
 
 	return 0;
 }
-device_initcall(apr_init);
+//device_initcall(apr_init);
 
-static int __init apr_late_init(void)
+//static int __init apr_late_init(void)
+int __init apr_late_init(void)
 {
 	int ret = 0;
 
@@ -1240,7 +1242,7 @@ static int __init apr_late_init(void)
 
 	return ret;
 }
-late_initcall(apr_late_init);
+//late_initcall(apr_late_init);
 
 static void __exit apr_exit(void)
 {
@@ -1251,12 +1253,13 @@ static void __exit apr_exit(void)
 __exitcall(apr_exit);
 
 #ifdef CONFIG_DEBUG_FS
-static int __init apr_debug_init(void)
+//static int __init apr_debug_init(void)
+int __init apr_debug_init(void)
 {
 	debugfs_apr_debug = debugfs_create_file("msm_apr_debug",
 						 S_IFREG | S_IRUGO, NULL, NULL,
 						 &apr_debug_ops);
 	return 0;
 }
-device_initcall(apr_debug_init);
+//device_initcall(apr_debug_init);
 #endif
diff --git a/drivers/soc/qcom/qdsp6v2/msm_audio_ion_vm.c b/drivers/soc/qcom/qdsp6v2/msm_audio_ion_vm.c
index c2415aa..6d84321 100644
--- a/drivers/soc/qcom/qdsp6v2/msm_audio_ion_vm.c
+++ b/drivers/soc/qcom/qdsp6v2/msm_audio_ion_vm.c
@@ -830,11 +830,12 @@ static struct platform_driver msm_audio_ion_driver = {
 	.remove = msm_audio_ion_remove,
 };
 
-static int __init msm_audio_ion_init(void)
+//static int __init msm_audio_ion_init(void)
+int __init msm_audio_ion_init(void)
 {
 	return platform_driver_register(&msm_audio_ion_driver);
 }
-module_init(msm_audio_ion_init);
+//module_init(msm_audio_ion_init);
 
 static void __exit msm_audio_ion_exit(void)
 {
diff --git a/drivers/soc/qcom/qdsp6v2/voice_svc.c b/drivers/soc/qcom/qdsp6v2/voice_svc.c
index 0a49a32..2001fd9 100644
--- a/drivers/soc/qcom/qdsp6v2/voice_svc.c
+++ b/drivers/soc/qcom/qdsp6v2/voice_svc.c
@@ -867,7 +867,8 @@ static struct platform_driver voice_svc_driver = {
 	},
 };
 
-static int __init voice_svc_init(void)
+//static int __init voice_svc_init(void)
+int __init voice_svc_init(void)
 {
 	pr_debug("%s\n", __func__);
 
@@ -881,7 +882,7 @@ static void __exit voice_svc_exit(void)
 	platform_driver_unregister(&voice_svc_driver);
 }
 
-module_init(voice_svc_init);
+//module_init(voice_svc_init);
 module_exit(voice_svc_exit);
 
 MODULE_DESCRIPTION("Soc QDSP6v2 Voice Service driver");
diff --git a/drivers/staging/android/ion/ion_hvenv_driver.c b/drivers/staging/android/ion/ion_hvenv_driver.c
index 5b9870d..87ab214 100644
--- a/drivers/staging/android/ion/ion_hvenv_driver.c
+++ b/drivers/staging/android/ion/ion_hvenv_driver.c
@@ -172,7 +172,8 @@ struct ion_client *hvenv_ion_client_create(const char *name)
 }
 EXPORT_SYMBOL(hvenv_ion_client_create);
 
-static int __init ion_dummy_init(void)
+//static int __init ion_dummy_init(void)
+int __init ion_dummy_init(void)
 {
 	int i, err;
 
@@ -225,7 +226,7 @@ err1:
 
 	return err;
 }
-device_initcall(ion_dummy_init);
+//device_initcall(ion_dummy_init);
 
 static void __exit ion_dummy_exit(void)
 {
diff --git a/init/do_mounts.c b/init/do_mounts.c
index 1902a1c..b6d7d43 100644
--- a/init/do_mounts.c
+++ b/init/do_mounts.c
@@ -563,7 +563,7 @@ void __init prepare_namespace(void)
 	 * For example, it is not atypical to wait 5 seconds here
 	 * for the touchpad of a laptop to initialize.
 	 */
-	wait_for_device_probe();
+	//wait_for_device_probe();
 
 	md_run_setup();
 	dm_run_setup();
@@ -599,6 +599,7 @@ void __init prepare_namespace(void)
 		ROOT_DEV = Root_RAM0;
 
 	mount_root();
+	wait_for_device_probe();
 out:
 	devtmpfs_mount("dev");
 	sys_mount(".", "/", NULL, MS_MOVE, NULL);
diff --git a/sound/core/compress_offload.c b/sound/core/compress_offload.c
index 15f58ea..81bbffb 100644
--- a/sound/core/compress_offload.c
+++ b/sound/core/compress_offload.c
@@ -1073,7 +1073,8 @@ int snd_compress_deregister(struct snd_compr *device)
 }
 EXPORT_SYMBOL_GPL(snd_compress_deregister);
 
-static int __init snd_compress_init(void)
+//static int __init snd_compress_init(void)
+int __init snd_compress_init(void)
 {
 	return 0;
 }
@@ -1082,7 +1083,7 @@ static void __exit snd_compress_exit(void)
 {
 }
 
-module_init(snd_compress_init);
+//module_init(snd_compress_init);
 module_exit(snd_compress_exit);
 
 MODULE_DESCRIPTION("ALSA Compressed offload framework");
diff --git a/sound/core/pcm.c b/sound/core/pcm.c
index 8239ebc..79348f3 100644
--- a/sound/core/pcm.c
+++ b/sound/core/pcm.c
@@ -1262,7 +1262,8 @@ static void snd_pcm_proc_done(void)
  *  ENTRY functions
  */
 
-static int __init alsa_pcm_init(void)
+//static int __init alsa_pcm_init(void)
+int __init alsa_pcm_init(void)
 {
 	snd_ctl_register_ioctl(snd_pcm_control_ioctl);
 	snd_ctl_register_ioctl_compat(snd_pcm_control_ioctl);
@@ -1277,5 +1278,5 @@ static void __exit alsa_pcm_exit(void)
 	snd_pcm_proc_done();
 }
 
-module_init(alsa_pcm_init)
+//module_init(alsa_pcm_init)
 module_exit(alsa_pcm_exit)
diff --git a/sound/core/rawmidi.c b/sound/core/rawmidi.c
index 89ae6ea..24449e5 100644
--- a/sound/core/rawmidi.c
+++ b/sound/core/rawmidi.c
@@ -1807,7 +1807,8 @@ EXPORT_SYMBOL(snd_rawmidi_set_ops);
  *  ENTRY functions
  */
 
-static int __init alsa_rawmidi_init(void)
+//static int __init alsa_rawmidi_init(void)
+int __init alsa_rawmidi_init(void)
 {
 
 	snd_ctl_register_ioctl(snd_rawmidi_control_ioctl);
@@ -1838,5 +1839,5 @@ static void __exit alsa_rawmidi_exit(void)
 	snd_ctl_unregister_ioctl_compat(snd_rawmidi_control_ioctl);
 }
 
-module_init(alsa_rawmidi_init)
+//module_init(alsa_rawmidi_init)
 module_exit(alsa_rawmidi_exit)
diff --git a/sound/core/timer.c b/sound/core/timer.c
index 94c563c..1b0865f 100644
--- a/sound/core/timer.c
+++ b/sound/core/timer.c
@@ -2073,7 +2073,8 @@ static struct device timer_dev;
  *  ENTRY functions
  */
 
-static int __init alsa_timer_init(void)
+//static int __init alsa_timer_init(void)
+int __init alsa_timer_init(void)
 {
 	int err;
 
@@ -2116,7 +2117,7 @@ static void __exit alsa_timer_exit(void)
 #endif
 }
 
-module_init(alsa_timer_init)
+//module_init(alsa_timer_init)
 module_exit(alsa_timer_exit)
 
 EXPORT_SYMBOL(snd_timer_open);
diff --git a/sound/last.c b/sound/last.c
index 43f2228..9dc17ac 100644
--- a/sound/last.c
+++ b/sound/last.c
@@ -22,7 +22,8 @@
 #include <linux/init.h>
 #include <sound/core.h>
 
-static int __init alsa_sound_last_init(void)
+//static int __init alsa_sound_last_init(void)
+int __init alsa_sound_last_init(void)
 {
 	int idx, ok = 0;
 	
@@ -37,4 +38,4 @@ static int __init alsa_sound_last_init(void)
 	return 0;
 }
 
-late_initcall_sync(alsa_sound_last_init);
+//late_initcall_sync(alsa_sound_last_init);
diff --git a/sound/soc/codecs/msm_stub.c b/sound/soc/codecs/msm_stub.c
index 9a9c54e..33a28d7 100755
--- a/sound/soc/codecs/msm_stub.c
+++ b/sound/soc/codecs/msm_stub.c
@@ -72,11 +72,12 @@ static struct platform_driver msm_stub_driver = {
 	.remove = msm_stub_dev_remove,
 };
 
-static int __init msm_stub_init(void)
+//static int __init msm_stub_init(void)
+int __init msm_stub_init(void)
 {
 	return platform_driver_register(&msm_stub_driver);
 }
-module_init(msm_stub_init);
+//module_init(msm_stub_init);
 
 static void __exit msm_stub_exit(void)
 {
diff --git a/sound/soc/msm/apq8096-auto.c b/sound/soc/msm/apq8096-auto.c
index 35fcd50..372ec27 100644
--- a/sound/soc/msm/apq8096-auto.c
+++ b/sound/soc/msm/apq8096-auto.c
@@ -23,6 +23,7 @@
 #include <linux/switch.h>
 #include <linux/input.h>
 #include <linux/qdsp6v2/apr.h>
+#include <linux/async.h>
 #include <sound/core.h>
 #include <sound/soc.h>
 #include <sound/soc-dapm.h>
@@ -7752,8 +7753,112 @@ static int __init apq8096_soc_platform_init(void)
 	return 0;
 }
 
-module_init(apq8096_soc_platform_init);
-module_platform_driver(apq8096_asoc_machine_driver);
+#ifdef CONFIG_DEBUG_FS
+extern int __init apr_debug_init(void);
+#endif
+extern int __init alsa_timer_init(void);
+extern int __init voice_init(void);
+extern int __init alsa_pcm_init(void);
+extern int __init alsa_rawmidi_init(void);
+extern int __init snd_compress_init(void);
+extern int __init snd_soc_init(void);
+extern int __init msm_stub_init(void);
+extern int __init msm_soc_platform_init_pcm_hostless(void);
+extern int __init msm_dai_slim_init(void);
+extern int __init audio_slimslave_init(void);
+extern int __init msm_dai_q6_init(void);
+extern int __init msm_soc_platform_init_pcm_q6_v2(void);
+extern int __init msm_soc_routing_platform_init(void);
+extern int __init msm_soc_platform_init_compress_q6_v2(void);
+extern int __init msm_soc_platform_init_pcm_afe_v2(void);
+extern int __init msm_soc_platform_init_pcm_voip_v2(void);
+extern int __init msm_soc_platform_init_pcm_voice_v2(void);
+extern int __init msm_dai_q6_hdmi_init(void);
+extern int __init msm_soc_platform_init_lsm_client(void);
+extern int __init msm_soc_platform_init_pcm_host_voice_v2(void);
+extern int __init msm_soc_platform_init_pcm_loopback_v2(void);
+extern int __init msm_soc_platform_init_transcode_loopback_q6_v2(void);
+extern int __init msm_soc_platform_init_pcm_dtmf_v2(void);
+extern int __init msm_dai_stub_init(void);
+extern int __init adm_init(void);
+extern int __init afe_init(void);
+extern int __init q6asm_init(void);
+extern int __init q6_core_init(void);
+extern int __init rtac_init(void);
+extern int __init q6lsm_init(void);
+extern int __init msm_soc_platform_init_pcm_q6_noirq(void);
+extern int __init msm_fe_dai_init(void);
+extern int __init apr_init(void);
+extern int __init alsa_sound_last_init(void);
+extern int __init apr_late_init(void);
+extern int __init voice_svc_init(void);
+extern int __init msm_audio_ion_init(void);
+extern int __init ion_dummy_init(void);
+
+static int __init apq8096_asoc_machine_driver_init(void)
+{
+	return platform_driver_register(&apq8096_asoc_machine_driver);
+}
+
+static void __exit apq8096_asoc_machine_driver_exit(void)
+{
+	return platform_driver_unregister(&apq8096_asoc_machine_driver);
+}
+module_exit(apq8096_asoc_machine_driver_exit);
+
+static void early_audio_init_async(void *data, async_cookie_t cookie)
+{
+	voice_init();
+	ion_dummy_init();
+#ifdef CONFIG_DEBUG_FS
+	apr_debug_init();
+#endif
+	apr_init();
+	voice_svc_init();
+	msm_audio_ion_init();
+	alsa_timer_init();
+	alsa_pcm_init();
+	alsa_rawmidi_init();
+	snd_compress_init();
+	snd_soc_init();
+	msm_stub_init();
+	msm_soc_platform_init_pcm_hostless();
+	msm_dai_slim_init();
+	audio_slimslave_init();
+	msm_dai_q6_init();
+	msm_soc_platform_init_pcm_q6_v2();
+	msm_soc_routing_platform_init();
+	msm_soc_platform_init_compress_q6_v2();
+	msm_soc_platform_init_pcm_afe_v2();
+	msm_soc_platform_init_pcm_voip_v2();
+	msm_soc_platform_init_pcm_voice_v2();
+	msm_dai_q6_hdmi_init();
+	msm_soc_platform_init_lsm_client();
+	msm_soc_platform_init_pcm_host_voice_v2();
+	msm_soc_platform_init_pcm_loopback_v2();
+	msm_soc_platform_init_transcode_loopback_q6_v2();
+	msm_soc_platform_init_pcm_dtmf_v2();
+	msm_dai_stub_init();
+	adm_init();
+	afe_init();
+	q6asm_init();
+	q6_core_init();
+	rtac_init();
+	q6lsm_init();
+	msm_soc_platform_init_pcm_q6_noirq();
+	msm_fe_dai_init();
+	apq8096_soc_platform_init();
+	apq8096_asoc_machine_driver_init();
+	alsa_sound_last_init();
+	apr_late_init();
+}
+
+int __init early_audio_init_async_start(void)
+{
+	async_schedule(early_audio_init_async, NULL);
+	return 0;
+}
+fs_initcall_sync(early_audio_init_async_start);
 
 MODULE_DESCRIPTION("ALSA SoC msm");
 MODULE_LICENSE("GPL v2");
diff --git a/sound/soc/msm/msm-dai-fe.c b/sound/soc/msm/msm-dai-fe.c
index c9178fb..df0217f 100644
--- a/sound/soc/msm/msm-dai-fe.c
+++ b/sound/soc/msm/msm-dai-fe.c
@@ -2872,11 +2872,12 @@ static struct platform_driver msm_fe_dai_driver = {
 	},
 };
 
-static int __init msm_fe_dai_init(void)
+//static int __init msm_fe_dai_init(void)
+int __init msm_fe_dai_init(void)
 {
 	return platform_driver_register(&msm_fe_dai_driver);
 }
-module_init(msm_fe_dai_init);
+//module_init(msm_fe_dai_init);
 
 static void __exit msm_fe_dai_exit(void)
 {
diff --git a/sound/soc/msm/msm-pcm-hostless.c b/sound/soc/msm/msm-pcm-hostless.c
index 51b0a72..d72ff4d 100644
--- a/sound/soc/msm/msm-pcm-hostless.c
+++ b/sound/soc/msm/msm-pcm-hostless.c
@@ -68,11 +68,12 @@ static struct platform_driver msm_pcm_hostless_driver = {
 	.remove = msm_pcm_hostless_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_hostless(void)
 {
 	return platform_driver_register(&msm_pcm_hostless_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/audio_slimslave.c b/sound/soc/msm/qdsp6v2/audio_slimslave.c
index e9ecfd5..fd1ad18 100644
--- a/sound/soc/msm/qdsp6v2/audio_slimslave.c
+++ b/sound/soc/msm/qdsp6v2/audio_slimslave.c
@@ -160,11 +160,12 @@ static struct slim_driver audio_slimslave_driver = {
 	.suspend = audio_slimslave_suspend,
 };
 
-static int __init audio_slimslave_init(void)
+//static int __init audio_slimslave_init(void)
+int __init audio_slimslave_init(void)
 {
 	return slim_driver_register(&audio_slimslave_driver);
 }
-module_init(audio_slimslave_init);
+//module_init(audio_slimslave_init);
 
 static void __exit audio_slimslave_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-compress-q6-v2.c b/sound/soc/msm/qdsp6v2/msm-compress-q6-v2.c
index 7a2c50c..d74edec 100644
--- a/sound/soc/msm/qdsp6v2/msm-compress-q6-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-compress-q6-v2.c
@@ -4977,11 +4977,12 @@ static struct platform_driver msm_compr_driver = {
 	.remove = msm_compr_dev_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_compress_q6_v2(void)
 {
 	return platform_driver_register(&msm_compr_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-dai-q6-hdmi-v2.c b/sound/soc/msm/qdsp6v2/msm-dai-q6-hdmi-v2.c
index 6b1c150..6d161fb 100644
--- a/sound/soc/msm/qdsp6v2/msm-dai-q6-hdmi-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-dai-q6-hdmi-v2.c
@@ -538,11 +538,12 @@ static struct platform_driver msm_dai_q6_hdmi_driver = {
 	},
 };
 
-static int __init msm_dai_q6_hdmi_init(void)
+//static int __init msm_dai_q6_hdmi_init(void)
+int __init msm_dai_q6_hdmi_init(void)
 {
 	return platform_driver_register(&msm_dai_q6_hdmi_driver);
 }
-module_init(msm_dai_q6_hdmi_init);
+//module_init(msm_dai_q6_hdmi_init);
 
 static void __exit msm_dai_q6_hdmi_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-dai-q6-v2.c b/sound/soc/msm/qdsp6v2/msm-dai-q6-v2.c
index 39abecb..d65b4c7 100644
--- a/sound/soc/msm/qdsp6v2/msm-dai-q6-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-dai-q6-v2.c
@@ -9642,7 +9642,8 @@ static struct platform_driver msm_dai_q6_tdm_driver = {
 	},
 };
 
-static int __init msm_dai_q6_init(void)
+//static int __init msm_dai_q6_init(void)
+int __init msm_dai_q6_init(void)
 {
 	int rc;
 
@@ -9731,7 +9732,7 @@ dai_q6_fail:
 fail:
 	return rc;
 }
-module_init(msm_dai_q6_init);
+//module_init(msm_dai_q6_init);
 
 static void __exit msm_dai_q6_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-dai-slim.c b/sound/soc/msm/qdsp6v2/msm-dai-slim.c
index 64ecd7e..7d1be0b 100644
--- a/sound/soc/msm/qdsp6v2/msm-dai-slim.c
+++ b/sound/soc/msm/qdsp6v2/msm-dai-slim.c
@@ -648,7 +648,8 @@ static struct slim_driver msm_dai_slim_driver = {
 	.id_table = msm_dai_slim_dt_match,
 };
 
-static int __init msm_dai_slim_init(void)
+//static int __init msm_dai_slim_init(void)
+int __init msm_dai_slim_init(void)
 {
 	int rc;
 	rc = slim_driver_register(&msm_dai_slim_driver);
@@ -657,7 +658,7 @@ static int __init msm_dai_slim_init(void)
 			__func__, rc);
 	return rc;
 }
-module_init(msm_dai_slim_init);
+//module_init(msm_dai_slim_init);
 
 static void __exit msm_dai_slim_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-dai-stub-v2.c b/sound/soc/msm/qdsp6v2/msm-dai-stub-v2.c
index 616a1b2..2cd9a98 100644
--- a/sound/soc/msm/qdsp6v2/msm-dai-stub-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-dai-stub-v2.c
@@ -354,7 +354,8 @@ static struct platform_driver msm_dai_stub_driver = {
 	},
 };
 
-static int __init msm_dai_stub_init(void)
+//static int __init msm_dai_stub_init(void)
+int __init msm_dai_stub_init(void)
 {
 	int rc = 0;
 
@@ -378,7 +379,7 @@ dai_stub_dev_fail:
 fail:
 	return rc;
 }
-module_init(msm_dai_stub_init);
+//module_init(msm_dai_stub_init);
 
 static void __exit msm_dai_stub_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-lsm-client.c b/sound/soc/msm/qdsp6v2/msm-lsm-client.c
index 9b0918c..2d9f1f1 100644
--- a/sound/soc/msm/qdsp6v2/msm-lsm-client.c
+++ b/sound/soc/msm/qdsp6v2/msm-lsm-client.c
@@ -2433,11 +2433,12 @@ static struct platform_driver msm_lsm_driver = {
 	.remove = msm_lsm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_lsm_client(void)
 {
 	return platform_driver_register(&msm_lsm_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-afe-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-afe-v2.c
index b1a1ea5..110cfa5 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-afe-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-afe-v2.c
@@ -906,12 +906,13 @@ static struct platform_driver msm_afe_driver = {
 	.remove = msm_afe_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_afe_v2(void)
 {
 	pr_debug("%s\n", __func__);
 	return platform_driver_register(&msm_afe_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-dtmf-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-dtmf-v2.c
index abbcfed..658882b 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-dtmf-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-dtmf-v2.c
@@ -580,11 +580,12 @@ static struct platform_driver msm_pcm_driver = {
 	.remove = msm_pcm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_dtmf_v2(void)
 {
 	return platform_driver_register(&msm_pcm_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-host-voice-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-host-voice-v2.c
index b02ab78..c94d5ca 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-host-voice-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-host-voice-v2.c
@@ -1503,7 +1503,8 @@ static struct platform_driver msm_pcm_driver = {
 	.remove = msm_pcm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_host_voice_v2(void)
 {
 	int i = 0;
 	struct session *s = NULL;
@@ -1540,7 +1541,7 @@ static int __init msm_soc_platform_init(void)
 
 	return platform_driver_register(&msm_pcm_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-loopback-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-loopback-v2.c
index 0add46f..fdd96c9 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-loopback-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-loopback-v2.c
@@ -1798,11 +1798,12 @@ static struct platform_driver msm_pcm_driver = {
 	.remove = msm_pcm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_loopback_v2(void)
 {
 	return platform_driver_register(&msm_pcm_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-q6-noirq.c b/sound/soc/msm/qdsp6v2/msm-pcm-q6-noirq.c
index 5c78349..82b4ef6 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-q6-noirq.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-q6-noirq.c
@@ -1319,11 +1319,12 @@ static struct platform_driver msm_pcm_driver_noirq = {
 	.remove = msm_pcm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_q6_noirq(void)
 {
 	return platform_driver_register(&msm_pcm_driver_noirq);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-q6-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-q6-v2.c
index b215462..84c8584 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-q6-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-q6-v2.c
@@ -3211,7 +3211,8 @@ static struct platform_driver msm_pcm_driver = {
 	.remove = msm_pcm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_q6_v2(void)
 {
 	init_waitqueue_head(&the_locks.enable_wait);
 	init_waitqueue_head(&the_locks.eos_wait);
@@ -3220,7 +3221,7 @@ static int __init msm_soc_platform_init(void)
 
 	return platform_driver_register(&msm_pcm_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-routing-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-routing-v2.c
index f0f57aa..9864603 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-routing-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-routing-v2.c
@@ -21221,7 +21221,8 @@ err:
 	return ret;
 }
 
-static int __init msm_soc_routing_platform_init(void)
+//static int __init msm_soc_routing_platform_init(void)
+int __init msm_soc_routing_platform_init(void)
 {
 	mutex_init(&routing_lock);
 	if (msm_routing_init_cal_data())
@@ -21235,7 +21236,7 @@ static int __init msm_soc_routing_platform_init(void)
 
 	return platform_driver_register(&msm_routing_pcm_driver);
 }
-module_init(msm_soc_routing_platform_init);
+//module_init(msm_soc_routing_platform_init);
 
 static void __exit msm_soc_routing_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-voice-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-voice-v2.c
index 2615ee6..6d53486 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-voice-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-voice-v2.c
@@ -1039,7 +1039,8 @@ static struct platform_driver msm_pcm_driver = {
 	.remove = msm_pcm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_voice_v2(void)
 {
 	int i = 0;
 
@@ -1050,7 +1051,7 @@ static int __init msm_soc_platform_init(void)
 
 	return platform_driver_register(&msm_pcm_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-pcm-voip-v2.c b/sound/soc/msm/qdsp6v2/msm-pcm-voip-v2.c
index b2387a7..bb644d7 100644
--- a/sound/soc/msm/qdsp6v2/msm-pcm-voip-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-pcm-voip-v2.c
@@ -1682,7 +1682,8 @@ static struct platform_driver msm_pcm_driver = {
 	.remove = msm_pcm_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_pcm_voip_v2(void)
 {
 	memset(&voip_info, 0, sizeof(voip_info));
 	voip_info.mode = MODE_PCM;
@@ -1701,7 +1702,7 @@ static int __init msm_soc_platform_init(void)
 
 	return platform_driver_register(&msm_pcm_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/msm-transcode-loopback-q6-v2.c b/sound/soc/msm/qdsp6v2/msm-transcode-loopback-q6-v2.c
index ef9c652..96453f2 100644
--- a/sound/soc/msm/qdsp6v2/msm-transcode-loopback-q6-v2.c
+++ b/sound/soc/msm/qdsp6v2/msm-transcode-loopback-q6-v2.c
@@ -1432,13 +1432,14 @@ static struct platform_driver msm_transcode_loopback_driver = {
 	.remove = msm_transcode_remove,
 };
 
-static int __init msm_soc_platform_init(void)
+//static int __init msm_soc_platform_init(void)
+int __init msm_soc_platform_init_transcode_loopback_q6_v2(void)
 {
 	memset(&transcode_info, 0, sizeof(struct msm_transcode_loopback));
 	mutex_init(&transcode_info.lock);
 	return platform_driver_register(&msm_transcode_loopback_driver);
 }
-module_init(msm_soc_platform_init);
+//module_init(msm_soc_platform_init);
 
 static void __exit msm_soc_platform_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/q6adm.c b/sound/soc/msm/qdsp6v2/q6adm.c
index dc71652..289928c 100644
--- a/sound/soc/msm/qdsp6v2/q6adm.c
+++ b/sound/soc/msm/qdsp6v2/q6adm.c
@@ -4701,7 +4701,8 @@ done:
 	return ret;
 }
 
-static int __init adm_init(void)
+//static int __init adm_init(void)
+int __init adm_init(void)
 {
 	int i = 0, j;
 
@@ -4731,5 +4732,5 @@ static void __exit adm_exit(void)
 	adm_delete_cal_data();
 }
 
-device_initcall(adm_init);
+//device_initcall(adm_init);
 module_exit(adm_exit);
diff --git a/sound/soc/msm/qdsp6v2/q6afe.c b/sound/soc/msm/qdsp6v2/q6afe.c
index 7ee774d..60573ae 100644
--- a/sound/soc/msm/qdsp6v2/q6afe.c
+++ b/sound/soc/msm/qdsp6v2/q6afe.c
@@ -7269,7 +7269,8 @@ done:
 }
 EXPORT_SYMBOL(afe_release_all_dma_resources);
 
-static int __init afe_init(void)
+//static int __init afe_init(void)
+int __init afe_init(void)
 {
 	int i = 0, ret;
 
@@ -7313,5 +7314,5 @@ static void __exit afe_exit(void)
 	wakeup_source_trash(&wl.ws);
 }
 
-device_initcall(afe_init);
+//device_initcall(afe_init);
 __exitcall(afe_exit);
diff --git a/sound/soc/msm/qdsp6v2/q6asm.c b/sound/soc/msm/qdsp6v2/q6asm.c
index 95374c0..f631fcc 100644
--- a/sound/soc/msm/qdsp6v2/q6asm.c
+++ b/sound/soc/msm/qdsp6v2/q6asm.c
@@ -10065,7 +10065,8 @@ static int q6asm_is_valid_session(struct apr_client_data *data, void *priv)
 	return 0;
 }
 
-static int __init q6asm_init(void)
+//static int __init q6asm_init(void)
+int __init q6asm_init(void)
 {
 	int lcnt, ret;
 	pr_debug("%s:\n", __func__);
@@ -10109,5 +10110,5 @@ static void __exit q6asm_exit(void)
 	q6asm_delete_cal_data();
 }
 
-device_initcall(q6asm_init);
+//device_initcall(q6asm_init);
 __exitcall(q6asm_exit);
diff --git a/sound/soc/msm/qdsp6v2/q6core.c b/sound/soc/msm/qdsp6v2/q6core.c
index 7ac3dcf..3e442ae 100644
--- a/sound/soc/msm/qdsp6v2/q6core.c
+++ b/sound/soc/msm/qdsp6v2/q6core.c
@@ -1249,7 +1249,8 @@ err:
 	return ret;
 }
 
-static int __init core_init(void)
+//static int __init core_init(void)
+int __init q6_core_init(void)
 {
 	memset(&q6core_lcl, 0, sizeof(struct q6core_str));
 	init_waitqueue_head(&q6core_lcl.bus_bw_req_wait);
@@ -1263,7 +1264,7 @@ static int __init core_init(void)
 
 	return 0;
 }
-module_init(core_init);
+//module_init(core_init);
 
 static void __exit core_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/q6lsm.c b/sound/soc/msm/qdsp6v2/q6lsm.c
index d926623..861bfae 100644
--- a/sound/soc/msm/qdsp6v2/q6lsm.c
+++ b/sound/soc/msm/qdsp6v2/q6lsm.c
@@ -2107,7 +2107,8 @@ err:
 	return ret;
 }
 
-static int __init q6lsm_init(void)
+//static int __init q6lsm_init(void)
+int __init q6lsm_init(void)
 {
 	int i = 0;
 	pr_debug("%s:\n", __func__);
@@ -2135,5 +2136,5 @@ static void __exit q6lsm_exit(void)
 	lsm_delete_cal_data();
 }
 
-device_initcall(q6lsm_init);
+//device_initcall(q6lsm_init);
 __exitcall(q6lsm_exit);
diff --git a/sound/soc/msm/qdsp6v2/q6voice.c b/sound/soc/msm/qdsp6v2/q6voice.c
index eb63524..917d443 100644
--- a/sound/soc/msm/qdsp6v2/q6voice.c
+++ b/sound/soc/msm/qdsp6v2/q6voice.c
@@ -8720,7 +8720,8 @@ int is_voc_initialized(void)
 	return module_initialized;
 }
 
-static int __init voice_init(void)
+//static int __init voice_init(void)
+int __init voice_init(void)
 {
 	int rc = 0, i = 0;
 
@@ -8816,7 +8817,7 @@ static int __init voice_init(void)
 	return rc;
 }
 
-device_initcall(voice_init);
+//device_initcall(voice_init);
 
 static void __exit voice_exit(void)
 {
diff --git a/sound/soc/msm/qdsp6v2/rtac.c b/sound/soc/msm/qdsp6v2/rtac.c
index 69e5464..7455d15 100644
--- a/sound/soc/msm/qdsp6v2/rtac.c
+++ b/sound/soc/msm/qdsp6v2/rtac.c
@@ -1987,7 +1987,8 @@ struct miscdevice rtac_misc = {
 	.fops	= &rtac_fops,
 };
 
-static int __init rtac_init(void)
+//static int __init rtac_init(void)
+int __init rtac_init(void)
 {
 	int i = 0;
 
@@ -2071,7 +2072,7 @@ nomem:
 	return -ENOMEM;
 }
 
-module_init(rtac_init);
+//module_init(rtac_init);
 
 MODULE_DESCRIPTION("SoC QDSP6v2 Real-Time Audio Calibration driver");
 MODULE_LICENSE("GPL v2");
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index 275df1d..cdc68e3 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -3765,14 +3765,15 @@ err:
 }
 EXPORT_SYMBOL_GPL(snd_soc_of_get_dai_link_codecs);
 
-static int __init snd_soc_init(void)
+//static int __init snd_soc_init(void)
+int __init snd_soc_init(void)
 {
 	snd_soc_debugfs_init();
 	snd_soc_util_init();
 
 	return platform_driver_register(&soc_driver);
 }
-module_init(snd_soc_init);
+//module_init(snd_soc_init);
 
 static void __exit snd_soc_exit(void)
 {
-- 
2.7.4

